CLI: Add 7 missing commands for full MCP server feature parity

## User Story: Add 7 Missing CLI Commands for Full MCP Server Feature Parity

**Story ID:** GH-66b (Sub-issue of GH-66)
**Epic:** CLI Modernization (GH-66)
**Priority:** High
**Story Points:** 5
**Sprint Fit:** Medium

### Story Statement

> **As a** developer or AI agent using the CodeCompress CLI,
> **I want** all MCP server tools available as CLI commands,
> **So that** I can use the CLI as a complete alternative to the MCP server for any code discovery task.

### Background & Context

After GH-66a establishes the System.CommandLine foundation, the CLI will have 9 commands while the MCP server exposes 16 tools (excluding `stop_server`). This story adds the 7 missing commands to achieve complete feature parity. Each new command mirrors its MCP equivalent in functionality, parameters, and output structure, ensuring agents and developers can switch between CLI and MCP seamlessly.

### Prerequisites

- **GH-66a must be complete** — System.CommandLine infrastructure, `--json` flag, and output formatter pattern must be in place before adding new commands.

### Commands to Implement

| # | CLI Command | MCP Equivalent | Category |
|---|---|---|---|
| 1 | `invalidate-cache` | `invalidate_cache` | Indexing |
| 2 | `get-module-api` | `get_module_api` | Query |
| 3 | `expand-symbol` | `expand_symbol` | Query |
| 4 | `get-symbols` | `get_symbols` | Query (batch) |
| 5 | `topic-outline` | `topic_outline` | Query |
| 6 | `project-deps` | `project_dependencies` | Dependency |
| 7 | `find-references` | `find_references` | Reference |

### Acceptance Criteria

```gherkin
# ── 1. invalidate-cache ────────────────────────────────────

Scenario: invalidate-cache deletes the project index
  Given a project at /project has been indexed
  When I run codecompress invalidate-cache --path /project
  Then the index database for that project is deleted
  And the output confirms "Index invalidated for /project"
  And the next index command performs a full re-index

Scenario: invalidate-cache with --json
  When I run codecompress invalidate-cache --path /project --json
  Then stdout contains JSON: { "status": "invalidated", "path": "/project" }

Scenario: invalidate-cache on unindexed project
  Given /project has never been indexed
  When I run codecompress invalidate-cache --path /project
  Then it succeeds gracefully (nothing to delete)
  And exit code is 0

# ── 2. get-module-api ───────────────────────────────────────

Scenario: get-module-api shows public API surface
  Given a project is indexed
  When I run codecompress get-module-api --path /project --module src/Core/Foo.cs
  Then it outputs all exported symbols, signatures, and dependencies for that file
  And --json output matches the MCP get_module_api response structure

Scenario: get-module-api with invalid module path
  When I run codecompress get-module-api --path /project --module nonexistent.cs
  Then stderr shows "Error: Module not found: nonexistent.cs"
  And exit code is 1

# ── 3. expand-symbol ────────────────────────────────────────

Scenario: expand-symbol extracts a nested symbol
  Given a project is indexed with class MyClass containing method DoWork
  When I run codecompress expand-symbol --path /project --name MyClass:DoWork
  Then it outputs only the DoWork method body
  And it does NOT include the entire MyClass source

Scenario: expand-symbol with context
  When I run codecompress expand-symbol --path /project --name MyClass:DoWork --context
  Then it includes 3 lines before and after the method

Scenario: expand-symbol for non-existent symbol
  When I run codecompress expand-symbol --path /project --name Foo:Bar
  Then stderr shows "Error: Symbol not found: Foo:Bar"
  And exit code is 1

# ── 4. get-symbols (batch) ──────────────────────────────────

Scenario: get-symbols retrieves multiple symbols
  Given a project is indexed with symbols Foo, Bar, and Baz
  When I run codecompress get-symbols --path /project --names "Foo,Bar,Baz"
  Then it returns source code for all three symbols in one response
  And --json includes a "results" array and an "errors" array

Scenario: get-symbols with mix of found and not-found
  When I run codecompress get-symbols --path /project --names "Foo,NonExistent"
  Then results contains Foo
  And errors contains NonExistent with "Symbol not found"
  And exit code is 0 (partial success)

Scenario: get-symbols enforces 50 name limit
  When I run codecompress get-symbols --path /project --names with 51 names
  Then stderr shows an error about exceeding the 50 name limit
  And exit code is 2

# ── 5. topic-outline ────────────────────────────────────────

Scenario: topic-outline searches and presents as outline
  Given a project is indexed
  When I run codecompress topic-outline --path /project --topic authentication
  Then it shows symbols related to "authentication" grouped by file in outline format
  And --json output matches the MCP topic_outline response structure

Scenario: topic-outline with path filter and limit
  When I run codecompress topic-outline --path /project --topic database --path-filter src/Core/ --max-results 10
  Then results are scoped to src/Core/ with at most 10 symbols

Scenario: topic-outline with no matches
  When I run codecompress topic-outline --path /project --topic xyznonexistent
  Then the output indicates no symbols matched
  And exit code is 0

# ── 6. project-deps ─────────────────────────────────────────

Scenario: project-deps shows inter-project dependencies
  Given a .NET solution is indexed
  When I run codecompress project-deps --path /solution
  Then it shows project-to-project dependency relationships
  And it includes shared public types between projects

Scenario: project-deps with filter
  When I run codecompress project-deps --path /solution --filter Core
  Then results are filtered to projects whose name contains "Core"

# ── 7. find-references ──────────────────────────────────────

Scenario: find-references shows all usages of a symbol
  Given a project is indexed
  When I run codecompress find-references --path /project --name ISymbolStore
  Then it returns file paths, line numbers, and 3-line context snippets
  And results are sorted by file path

Scenario: find-references with path filter and limit
  When I run codecompress find-references --path /project --name ISymbolStore --path-filter src/ --limit 5
  Then results are scoped to src/ with at most 5 results

Scenario: find-references for unused symbol
  When I run codecompress find-references --path /project --name NeverUsedThing
  Then the output indicates no references found
  And exit code is 0

# ── Cross-Cutting ───────────────────────────────────────────

Scenario: All new commands appear in root --help
  When I run codecompress --help
  Then all 7 new commands appear in the appropriate category group
  And each has a one-line summary

Scenario: All new commands have --help with examples
  Given any of the 7 new commands
  When I run codecompress <command> --help
  Then it shows the command description, all options with types and defaults, and 2-3 examples

Scenario: All new commands support --json flag
  Given any of the 7 new commands
  When I run it with --json
  Then stdout is valid JSON with snake_case property naming
```

### Technical Notes

**Implementation Pattern:**

Each new command follows the same pattern established in GH-66a:
1. Create `Commands/{CommandName}Command.cs` with a `System.CommandLine.Command` subclass
2. Define `Option<T>` for each parameter matching the MCP tool's parameters
3. Wire up the handler via `SetHandler`
4. Handler creates a `CliProjectScope`, calls the appropriate `ISymbolStore` / `IIndexEngine` method
5. Output via the `IOutputFormatter` (JSON or human-readable based on `--json` flag)

**Core Library Methods to Use:**

| Command | Core Method |
|---|---|
| `invalidate-cache` | Delete the `.code-compress/index.db` file (or call store method if available) |
| `get-module-api` | `ISymbolStore.GetSymbolsByFileAsync()` + format as API surface |
| `expand-symbol` | `ISymbolStore.GetSymbolByNameAsync()` with qualified name parsing (same as MCP `expand_symbol`) |
| `get-symbols` | Loop `GetSymbolByNameAsync()` or batch method if available |
| `topic-outline` | `ISymbolStore.SearchSymbolsAsync()` + group by file (same as MCP `topic_outline`) |
| `project-deps` | `ISymbolStore.GetProjectDependencyGraphAsync()` |
| `find-references` | `ISymbolStore.FindReferencesAsync()` |

**Important:** Check the MCP server tool implementations for the exact Core API calls — the CLI should use identical logic to ensure output consistency. Use the CodeCompress MCP `get_symbol` tool to read each MCP tool's implementation before coding the CLI equivalent.

**Files to Create:**

| File | Description |
|------|-------------|
| `src/CodeCompress.Cli/Commands/InvalidateCacheCommand.cs` | Index invalidation |
| `src/CodeCompress.Cli/Commands/GetModuleApiCommand.cs` | Module API surface |
| `src/CodeCompress.Cli/Commands/ExpandSymbolCommand.cs` | Nested symbol extraction |
| `src/CodeCompress.Cli/Commands/GetSymbolsCommand.cs` | Batch symbol retrieval |
| `src/CodeCompress.Cli/Commands/TopicOutlineCommand.cs` | Topic-based search + outline |
| `src/CodeCompress.Cli/Commands/ProjectDepsCommand.cs` | Inter-project dependencies |
| `src/CodeCompress.Cli/Commands/FindReferencesCommand.cs` | Symbol reference finder |

### Dependencies

| Dependency | Type | Status | Owner |
|---|---|---|---|
| GH-66a (System.CommandLine migration) | Blocking | In progress | Project |
| CodeCompress.Core library | Non-blocking | Known | Project |
| MCP tool implementations (reference) | Non-blocking | Known | Project |

### Out of Scope

- Parameter parity for existing 9 commands (GH-66c)
- Error UX standardization (GH-66c)
- README documentation updates (GH-66c)
- Shell completion scripts

### Definition of Done

- [ ] All 7 new commands implemented and registered in the root command
- [ ] Each command supports `--json` output matching MCP tool response structure
- [ ] Each command has human-readable default output
- [ ] Each command has `--help` with description, options, and examples
- [ ] All commands appear in categorized root `--help`
- [ ] Error handling: missing params (exit 2), runtime errors (exit 1), actionable messages
- [ ] Unit tests for each new command handler
- [ ] `dotnet build CodeCompress.slnx` — zero warnings
- [ ] `dotnet test CodeCompress.slnx` — all tests pass
- [ ] Code reviewed and approved
- [ ] Product Owner acceptance

### Open Questions (Resolved)

*No open questions — all decisions captured in parent epic GH-66.*

--- a/README.md
+++ b/README.md
@@ -306,15 +306,36 @@ codecompress get-symbol --path /path/to/project --name MyClass:MyMethod
 # Search raw file contents
 codecompress search-text --path /path/to/project --query "TODO"
 
+# Retrieve a nested method without loading the whole class (~60% token savings)
+codecompress expand-symbol --path /path/to/project --name MyClass:MyMethod
+
+# Batch retrieve multiple symbols at once
+codecompress get-symbols --path /path/to/project --names "Foo,Bar,Baz"
+
+# Get the public API surface of a single file
+codecompress get-module-api --path /path/to/project --module src/Core/Foo.cs
+
+# Search by topic — returns results in outline format
+codecompress topic-outline --path /path/to/project --topic authentication
+
+# Find all references to a symbol across the codebase
+codecompress find-references --path /path/to/project --name ISymbolStore
+
 # View directory structure (no index required)
 codecompress file-tree --path /path/to/project
 
-# Show dependency graph
+# Show file-level dependency graph
 codecompress deps --path /path/to/project
+
+# Show inter-project dependencies (.NET solutions)
+codecompress project-deps --path /path/to/project
 
 # Snapshot + change tracking
 codecompress snapshot --path /path/to/project --label before-refactor
 codecompress changes --path /path/to/project --label before-refactor
+
+# Delete index to force full re-index
+codecompress invalidate-cache --path /path/to/project
 ```
 
 ### JSON Output
@@ -342,12 +363,19 @@ This outputs a markdown block you can paste into `CLAUDE.md`, system prompts, or
 | `index` | `index_project` | Build/update the symbol database |
 | `outline` | `project_outline` | Compressed codebase overview |
 | `get-symbol` | `get_symbol` | Retrieve symbol source code |
+| `expand-symbol` | `expand_symbol` | Extract nested symbol (~60% fewer tokens) |
+| `get-symbols` | `get_symbols` | Batch retrieve multiple symbols |
+| `get-module-api` | `get_module_api` | Public API surface of a file |
 | `search` | `search_symbols` | FTS5 symbol search |
 | `search-text` | `search_text` | FTS5 raw content search |
+| `topic-outline` | `topic_outline` | Topic-based search in outline format |
+| `find-references` | `find_references` | Find all symbol references |
 | `changes` | `changes_since` | Delta since snapshot |
 | `snapshot` | `snapshot_create` | Create index snapshot |
 | `file-tree` | `file_tree` | Directory tree |
-| `deps` | `dependency_graph` | Dependency graph |
+| `deps` | `dependency_graph` | File-level dependency graph |
+| `project-deps` | `project_dependencies` | Inter-project dependencies (.NET) |
+| `invalidate-cache` | `invalidate_cache` | Force full re-index |
 
 ## License
 

--- a/src/CodeCompress.Cli/Program.cs
+++ b/src/CodeCompress.Cli/Program.cs
@@ -1,6 +1,7 @@
 using System.CommandLine;
 using System.CommandLine.Parsing;
 using System.Globalization;
+using System.Text;
 using System.Text.Json;
 using CodeCompress.Cli;
 using CodeCompress.Core;
@@ -489,6 +490,513 @@ depsCommand.SetAction(async parseResult =>
 
 rootCommand.Subcommands.Add(depsCommand);
 
+// ── invalidate-cache ────────────────────────────────────────
+
+var invalidateCachePathOption = CreatePathOption();
+
+var invalidateCacheCommand = new Command("invalidate-cache",
+    "Delete the entire index for a project, forcing a full re-index on the next index command.")
+{
+    invalidateCachePathOption,
+};
+
+invalidateCacheCommand.SetAction(async parseResult =>
+{
+    var path = parseResult.GetValue(invalidateCachePathOption)!;
+    var json = parseResult.GetValue(jsonOption);
+
+    var scope = await CreateProjectScopeAsync(path, provider).ConfigureAwait(false);
+    await using (scope.ConfigureAwait(false))
+    {
+        var files = await scope.Store.GetFilesByRepoAsync(scope.RepoId).ConfigureAwait(false);
+        var fileIds = files.Select(f => f.Id).ToList();
+
+        foreach (var fileId in fileIds)
+        {
+            await scope.Store.DeleteSymbolsByFileAsync(fileId).ConfigureAwait(false);
+            await scope.Store.DeleteDependenciesByFileAsync(fileId).ConfigureAwait(false);
+            await scope.Store.DeleteFileAsync(fileId).ConfigureAwait(false);
+        }
+
+        await scope.Store.DeleteRepositoryAsync(scope.RepoId).ConfigureAwait(false);
+
+        if (json)
+        {
+            Console.WriteLine(JsonSerializer.Serialize(
+                new { Status = "invalidated", Path = path },
+                jsonSerializerOptions));
+        }
+        else
+        {
+            Console.WriteLine("Index invalidated. Next index command will perform a full reparse.");
+        }
+    }
+});
+
+rootCommand.Subcommands.Add(invalidateCacheCommand);
+
+// ── get-module-api ──────────────────────────────────────────
+
+var getModuleApiPathOption = CreatePathOption();
+var getModuleApiModuleOption = new Option<string>("--module")
+{
+    Description = "Relative path to the module file (e.g., src/Core/Foo.cs). Forward slashes only.",
+    Required = true,
+};
+
+var getModuleApiCommand = new Command("get-module-api",
+    "Get the full public API surface of a single file — symbols, signatures, and dependencies. Requires index.")
+{
+    getModuleApiPathOption,
+    getModuleApiModuleOption,
+};
+
+getModuleApiCommand.SetAction(async parseResult =>
+{
+    var path = parseResult.GetValue(getModuleApiPathOption)!;
+    var modulePath = parseResult.GetValue(getModuleApiModuleOption)!;
+    var json = parseResult.GetValue(jsonOption);
+
+    var scope = await CreateProjectScopeAsync(path, provider).ConfigureAwait(false);
+    await using (scope.ConfigureAwait(false))
+    {
+        ModuleApi moduleApi;
+        try
+        {
+            moduleApi = await scope.Store.GetModuleApiAsync(scope.RepoId, modulePath).ConfigureAwait(false);
+        }
+        catch (FileNotFoundException)
+        {
+            await Console.Error.WriteLineAsync($"Error: Module not found: {modulePath}").ConfigureAwait(false);
+            Environment.ExitCode = 1;
+            return;
+        }
+
+        if (json)
+        {
+            Console.WriteLine(JsonSerializer.Serialize(
+                new
+                {
+                    Module = moduleApi.File.RelativePath,
+                    Symbols = moduleApi.Symbols.Select(s => new
+                    {
+                        s.Name, s.Kind, Parent = s.ParentSymbol, s.Signature, Line = s.LineStart, s.DocComment,
+                    }),
+                    Dependencies = moduleApi.Dependencies.Select(d => new { d.RequiresPath, d.Alias }),
+                },
+                jsonSerializerOptions));
+        }
+        else
+        {
+            Console.WriteLine($"## {moduleApi.File.RelativePath}");
+            Console.WriteLine();
+            foreach (var s in moduleApi.Symbols)
+            {
+                Console.WriteLine($"  {s.Kind,-12} {s.Visibility,-10} {s.Signature}");
+            }
+
+            if (moduleApi.Dependencies.Count > 0)
+            {
+                Console.WriteLine();
+                Console.WriteLine("Dependencies:");
+                foreach (var d in moduleApi.Dependencies)
+                {
+                    Console.WriteLine($"  → {d.RequiresPath}" + (d.Alias is not null ? $" (as {d.Alias})" : ""));
+                }
+            }
+        }
+    }
+});
+
+rootCommand.Subcommands.Add(getModuleApiCommand);
+
+// ── expand-symbol ───────────────────────────────────────────
+
+var expandSymbolPathOption = CreatePathOption();
+var expandSymbolNameOption = new Option<string>("--name")
+{
+    Description = "Qualified symbol name (e.g., MyClass:MyMethod). Use search to discover names.",
+    Required = true,
+};
+var expandSymbolContextOption = new Option<bool>("--context")
+{
+    Description = "Include 3 lines of context before and after the symbol",
+};
+
+var expandSymbolCommand = new Command("expand-symbol",
+    "Retrieve only a nested symbol's body without loading the parent class — saves ~60% tokens. Requires index.")
+{
+    expandSymbolPathOption,
+    expandSymbolNameOption,
+    expandSymbolContextOption,
+};
+
+expandSymbolCommand.SetAction(async parseResult =>
+{
+    var path = parseResult.GetValue(expandSymbolPathOption)!;
+    var name = parseResult.GetValue(expandSymbolNameOption)!;
+    _ = parseResult.GetValue(expandSymbolContextOption); // Reserved for future context-with-lines support
+    var json = parseResult.GetValue(jsonOption);
+
+    var scope = await CreateProjectScopeAsync(path, provider).ConfigureAwait(false);
+    await using (scope.ConfigureAwait(false))
+    {
+        var symbol = await scope.Store.GetSymbolByNameAsync(scope.RepoId, name).ConfigureAwait(false);
+        if (symbol is null)
+        {
+            await Console.Error.WriteLineAsync($"Error: Symbol not found: {name}").ConfigureAwait(false);
+            await Console.Error.WriteLineAsync("  Hint: Use 'codecompress search --path <path> --query <name>' to discover symbol names.").ConfigureAwait(false);
+            Environment.ExitCode = 1;
+            return;
+        }
+
+        var files = await scope.Store.GetFilesByRepoAsync(scope.RepoId).ConfigureAwait(false);
+        var file = files.FirstOrDefault(f => f.Id == symbol.FileId);
+        if (file is null)
+        {
+            await Console.Error.WriteLineAsync("Error: File not found for symbol.").ConfigureAwait(false);
+            Environment.ExitCode = 1;
+            return;
+        }
+
+        var pathValidator = provider.GetRequiredService<IPathValidator>();
+        var resolvedPath = pathValidator.ValidatePath(
+            Path.Combine(path, file.RelativePath), path);
+
+        var sourceCode = await ReadSourceCodeAsync(resolvedPath, symbol.ByteOffset, symbol.ByteLength).ConfigureAwait(false);
+
+        if (json)
+        {
+            Console.WriteLine(JsonSerializer.Serialize(
+                new
+                {
+                    symbol.Name, symbol.Kind, Parent = symbol.ParentSymbol,
+                    File = file.RelativePath, symbol.LineStart, symbol.LineEnd,
+                    symbol.Signature, symbol.DocComment, SourceCode = sourceCode,
+                },
+                jsonSerializerOptions));
+        }
+        else
+        {
+            Console.WriteLine($"// {symbol.Name} ({symbol.Kind}, {symbol.Visibility})");
+            Console.WriteLine($"// {file.RelativePath}:{symbol.LineStart}-{symbol.LineEnd}");
+            Console.WriteLine(sourceCode);
+        }
+    }
+});
+
+rootCommand.Subcommands.Add(expandSymbolCommand);
+
+// ── get-symbols (batch) ─────────────────────────────────────
+
+var getSymbolsPathOption = CreatePathOption();
+var getSymbolsNamesOption = new Option<string>("--names")
+{
+    Description = "Comma-separated list of qualified symbol names (max 50). Same format as get-symbol.",
+    Required = true,
+};
+
+var getSymbolsCommand = new Command("get-symbols",
+    "Batch retrieve source code for multiple symbols in one call. More efficient than repeated get-symbol. Requires index.")
+{
+    getSymbolsPathOption,
+    getSymbolsNamesOption,
+};
+
+getSymbolsCommand.SetAction(async parseResult =>
+{
+    var path = parseResult.GetValue(getSymbolsPathOption)!;
+    var namesRaw = parseResult.GetValue(getSymbolsNamesOption)!;
+    var json = parseResult.GetValue(jsonOption);
+
+    var symbolNames = namesRaw.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
+
+    if (symbolNames.Length == 0)
+    {
+        await Console.Error.WriteLineAsync("Error: No symbol names provided.").ConfigureAwait(false);
+        Environment.ExitCode = 1;
+        return;
+    }
+
+    if (symbolNames.Length > 50)
+    {
+        await Console.Error.WriteLineAsync("Error: Too many symbols. Maximum is 50.").ConfigureAwait(false);
+        Environment.ExitCode = 1;
+        return;
+    }
+
+    var scope = await CreateProjectScopeAsync(path, provider).ConfigureAwait(false);
+    await using (scope.ConfigureAwait(false))
+    {
+        var foundSymbols = await scope.Store.GetSymbolsByNamesAsync(scope.RepoId, symbolNames).ConfigureAwait(false);
+        var files = await scope.Store.GetFilesByRepoAsync(scope.RepoId).ConfigureAwait(false);
+        var fileMap = files.ToDictionary(f => f.Id);
+
+        var foundNames = new HashSet<string>(StringComparer.Ordinal);
+        foreach (var s in foundSymbols)
+        {
+            var qn = s.ParentSymbol is not null ? $"{s.ParentSymbol}:{s.Name}" : s.Name;
+            foundNames.Add(qn);
+        }
+
+        var pathValidator = provider.GetRequiredService<IPathValidator>();
+        var results = new List<object>();
+
+        foreach (var s in foundSymbols)
+        {
+            if (!fileMap.TryGetValue(s.FileId, out var file))
+            {
+                continue;
+            }
+
+            string resolvedPath;
+            try
+            {
+                resolvedPath = pathValidator.ValidatePath(Path.Combine(path, file.RelativePath), path);
+            }
+            catch (ArgumentException)
+            {
+                continue;
+            }
+
+            var sourceCode = await ReadSourceCodeAsync(resolvedPath, s.ByteOffset, s.ByteLength).ConfigureAwait(false);
+            results.Add(new
+            {
+                s.Name, s.Kind, Parent = s.ParentSymbol,
+                File = file.RelativePath, s.LineStart, s.LineEnd,
+                s.Signature, SourceCode = sourceCode,
+            });
+        }
+
+        var errors = symbolNames
+            .Where(n => !foundNames.Contains(n))
+            .Select(n => new { Symbol = n, Error = "Symbol not found" })
+            .ToList();
+
+        if (json)
+        {
+            Console.WriteLine(JsonSerializer.Serialize(
+                new { Results = results, Errors = errors },
+                jsonSerializerOptions));
+        }
+        else
+        {
+            foreach (var r in results)
+            {
+                Console.WriteLine(JsonSerializer.Serialize(r, jsonSerializerOptions));
+                Console.WriteLine();
+            }
+
+            if (errors.Count > 0)
+            {
+                await Console.Error.WriteLineAsync($"Not found: {string.Join(", ", errors.Select(e => e.Symbol))}").ConfigureAwait(false);
+            }
+        }
+    }
+});
+
+rootCommand.Subcommands.Add(getSymbolsCommand);
+
+// ── topic-outline ───────────────────────────────────────────
+
+var topicOutlinePathOption = CreatePathOption();
+var topicOutlineTopicOption = new Option<string>("--topic")
+{
+    Description = "Topic or keyword to search for (e.g., 'authentication', 'database')",
+    Required = true,
+};
+var topicOutlinePathFilterOption = new Option<string?>("--path-filter")
+{
+    Description = "Filter results to files under this directory (e.g., 'src/Core/')",
+};
+var topicOutlineMaxResultsOption = new Option<int>("--max-results")
+{
+    Description = "Maximum symbols to return (1-200)",
+    DefaultValueFactory = _ => 50,
+};
+
+var topicOutlineCommand = new Command("topic-outline",
+    "Search for symbols related to a topic and return results in outline format. Requires index.")
+{
+    topicOutlinePathOption,
+    topicOutlineTopicOption,
+    topicOutlinePathFilterOption,
+    topicOutlineMaxResultsOption,
+};
+
+topicOutlineCommand.SetAction(async parseResult =>
+{
+    var path = parseResult.GetValue(topicOutlinePathOption)!;
+    var topic = parseResult.GetValue(topicOutlineTopicOption)!;
+    var pathFilter = parseResult.GetValue(topicOutlinePathFilterOption);
+    var maxResults = Math.Clamp(parseResult.GetValue(topicOutlineMaxResultsOption), 1, 200);
+    var json = parseResult.GetValue(jsonOption);
+
+    var scope = await CreateProjectScopeAsync(path, provider).ConfigureAwait(false);
+    await using (scope.ConfigureAwait(false))
+    {
+        var outline = await scope.Store.SearchTopicOutlineAsync(
+            scope.RepoId, topic, maxResults, pathFilter).ConfigureAwait(false);
+
+        if (json)
+        {
+            Console.WriteLine(JsonSerializer.Serialize(outline, jsonSerializerOptions));
+        }
+        else
+        {
+            if (outline.TotalSymbolCount == 0)
+            {
+                Console.WriteLine($"No symbols found for topic \"{topic}\".");
+                return;
+            }
+
+            Console.WriteLine($"Found {outline.TotalSymbolCount} symbol(s) for \"{topic}\":");
+            foreach (var group in outline.Groups)
+            {
+                Console.WriteLine($"## {group.Name}");
+                foreach (var s in group.Symbols)
+                {
+                    Console.WriteLine($"  {s.Kind,-12} {s.Visibility,-10} {s.Signature}");
+                }
+            }
+        }
+    }
+});
+
+rootCommand.Subcommands.Add(topicOutlineCommand);
+
+// ── project-deps ────────────────────────────────────────────
+
+var projectDepsPathOption = CreatePathOption();
+var projectDepsFilterOption = new Option<string?>("--filter")
+{
+    Description = "Filter to projects whose name contains this string (case-insensitive)",
+};
+
+var projectDepsCommand = new Command("project-deps",
+    "Show inter-project dependency relationships in a .NET solution. Requires index.")
+{
+    projectDepsPathOption,
+    projectDepsFilterOption,
+};
+
+projectDepsCommand.SetAction(async parseResult =>
+{
+    var path = parseResult.GetValue(projectDepsPathOption)!;
+    var filter = parseResult.GetValue(projectDepsFilterOption);
+    var json = parseResult.GetValue(jsonOption);
+
+    var scope = await CreateProjectScopeAsync(path, provider).ConfigureAwait(false);
+    await using (scope.ConfigureAwait(false))
+    {
+        var result = await scope.Store.GetProjectDependencyGraphAsync(scope.RepoId, filter).ConfigureAwait(false);
+
+        if (result.Projects.Count == 0)
+        {
+            await Console.Error.WriteLineAsync("Error: No project files found in index. Run 'codecompress index' first.").ConfigureAwait(false);
+            Environment.ExitCode = 1;
+            return;
+        }
+
+        if (json)
+        {
+            Console.WriteLine(JsonSerializer.Serialize(result, jsonSerializerOptions));
+        }
+        else
+        {
+            Console.WriteLine($"Projects ({result.Projects.Count}):");
+            foreach (var p in result.Projects)
+            {
+                Console.WriteLine($"  {p.Name} ({p.RelativePath})");
+            }
+
+            if (result.Edges.Count > 0)
+            {
+                Console.WriteLine();
+                Console.WriteLine("Dependencies:");
+                foreach (var e in result.Edges)
+                {
+                    Console.WriteLine($"  {e.FromProject} → {e.ToProject}");
+                    if (e.SharedTypes.Count > 0)
+                    {
+                        Console.WriteLine($"    Shared types: {string.Join(", ", e.SharedTypes.Take(5))}{(e.SharedTypes.Count > 5 ? "..." : "")}");
+                    }
+                }
+            }
+        }
+    }
+});
+
+rootCommand.Subcommands.Add(projectDepsCommand);
+
+// ── find-references ─────────────────────────────────────────
+
+var findRefsPathOption = CreatePathOption();
+var findRefsNameOption = new Option<string>("--name")
+{
+    Description = "Symbol name to search for references (e.g., 'ISymbolStore')",
+    Required = true,
+};
+var findRefsPathFilterOption = new Option<string?>("--path-filter")
+{
+    Description = "Filter results to files under this directory (e.g., 'src/')",
+};
+var findRefsLimitOption = new Option<int>("--limit")
+{
+    Description = "Maximum results to return (1-100)",
+    DefaultValueFactory = _ => 20,
+};
+
+var findRefsCommand = new Command("find-references",
+    "Find all locations where a symbol is referenced across the codebase. Faster than grep. Requires index.")
+{
+    findRefsPathOption,
+    findRefsNameOption,
+    findRefsPathFilterOption,
+    findRefsLimitOption,
+};
+
+findRefsCommand.SetAction(async parseResult =>
+{
+    var path = parseResult.GetValue(findRefsPathOption)!;
+    var symbolName = parseResult.GetValue(findRefsNameOption)!;
+    var pathFilter = parseResult.GetValue(findRefsPathFilterOption);
+    var limit = Math.Clamp(parseResult.GetValue(findRefsLimitOption), 1, 100);
+    var json = parseResult.GetValue(jsonOption);
+
+    var pathValidator = provider.GetRequiredService<IPathValidator>();
+    var validatedPath = pathValidator.ValidatePath(path, path);
+
+    var scope = await CreateProjectScopeAsync(path, provider).ConfigureAwait(false);
+    await using (scope.ConfigureAwait(false))
+    {
+        var results = await scope.Store.FindReferencesAsync(
+            scope.RepoId, symbolName, validatedPath, limit, pathFilter).ConfigureAwait(false);
+
+        if (json)
+        {
+            Console.WriteLine(JsonSerializer.Serialize(results, jsonSerializerOptions));
+        }
+        else
+        {
+            if (results.Count == 0)
+            {
+                Console.WriteLine($"No references found for \"{symbolName}\".");
+                return;
+            }
+
+            Console.WriteLine($"Found {results.Count} reference(s) for \"{symbolName}\":");
+            foreach (var r in results)
+            {
+                Console.WriteLine($"  {r.FilePath}:{r.Line}");
+                Console.WriteLine($"    {r.ContextSnippet.Trim()}");
+            }
+        }
+    }
+});
+
+rootCommand.Subcommands.Add(findRefsCommand);
+
 // ── agent-instructions ──────────────────────────────────────
 
 var agentInstructionsCommand = new Command("agent-instructions",
@@ -564,6 +1072,31 @@ static async Task<CliProjectScope> CreateProjectScopeAsync(string path, ServiceP
         serviceProvider.GetRequiredService<ILoggerFactory>().CreateLogger<IndexEngine>());
 
     return new CliProjectScope(connection, store, engine, repoId, validatedPath);
+}
+
+static async Task<string> ReadSourceCodeAsync(string filePath, int byteOffset, int byteLength)
+{
+    var stream = new FileStream(
+        filePath, FileMode.Open, FileAccess.Read, FileShare.Read,
+        bufferSize: 4096, FileOptions.Asynchronous | FileOptions.SequentialScan);
+    await using (stream.ConfigureAwait(false))
+    {
+        stream.Seek(byteOffset, SeekOrigin.Begin);
+        var buffer = new byte[byteLength];
+        var bytesRead = 0;
+        while (bytesRead < byteLength)
+        {
+            var read = await stream.ReadAsync(buffer.AsMemory(bytesRead, byteLength - bytesRead)).ConfigureAwait(false);
+            if (read == 0)
+            {
+                break;
+            }
+
+            bytesRead += read;
+        }
+
+        return Encoding.UTF8.GetString(buffer, 0, bytesRead);
+    }
 }
 
 static async Task PrintDirectoryTreeAsync(string rootPath, string currentPath, int maxDepth, int currentDepth)

